chainimport/headers_import.go

```diff
@@ -564,75 +764,111 @@ func (h *headersImport) appendNewHeaders(ctx context.Context, startHeight,
-	)
-	if filterErr == io.EOF {
-		return 0, errors.New("filter headers ended before block " +
-			"headers")
-	}
-	if filterErr != nil {
-		return 0, fmt.Errorf("failed to read filter headers "+
-			"batch at height %d: %w", batchStart, filterErr)
-	}
+	if appendMode != appendFilterOnly {
+		blockBatch, blockErr := blockIter.ReadBatch(
+			batchStart, blockIter.GetEndIndex(),
+			blockIter.GetBatchSize(),
+		)
+		if blockErr == io.EOF {
+			return 0, io.EOF
+		}
+		if blockErr != nil {
+			return 0, fmt.Errorf("failed to read block headers "+
+				"batch at height %d: %w", batchStart, blockErr)
+		}
 
-	// Convert block header batches to target store format.
-	blockHeaders := make([]headerfs.BlockHeader, 0, len(blockBatch))
-	for _, header := range blockBatch {
-		blkHeader, err := assertBlockHeader(header)
-		if err != nil {
-			return 0, err
+		// Convert block header batches to target store format.
+		blockHeaders = make([]headerfs.BlockHeader, 0, len(blockBatch))
+		for _, header := range blockBatch {
+			blkHeader, err := assertBlockHeader(header)
+			if err != nil {
+				return 0, err
+			}
+			blockHeaders = append(
+				blockHeaders, blkHeader.BlockHeader,
+			)
 		}
-		blockHeaders = append(
-			blockHeaders, blkHeader.BlockHeader,
-		)
+
+		batchEnd = batchStart + uint32(len(blockBatch)) - 1
 	}
 
-	// Convert filter header batches to target store format.
-	var filterHeaders []headerfs.FilterHeader
-	filterHeaders = make(
-		[]headerfs.FilterHeader, 0, len(filterBatch),
-	)
-	for _, header := range filterBatch {
-		fHeader, err := assertFilterHeader(header)
-		if err != nil {
-			return 0, err
-		}
-		filterHeaders = append(
-			filterHeaders, fHeader.FilterHeader,
+	if appendMode != appendBlockOnly {
+		filterBatch, filterErr := filterIter.ReadBatch(
+			batchStart, blockIter.GetEndIndex(),
+			blockIter.GetBatchSize(),
 		)
-	}
+		if filterErr == io.EOF {
+			return 0, io.EOF
+		}
+		if filterErr != nil {
+			return 0, fmt.Errorf("failed to read filter headers "+
+				"batch at height %d: %w", batchStart, filterErr)
+		}
 
-	// The length check conditions should never be triggered during normal
-	// import operations as validation occurs earlier. They serve as sanity
-	// checks to catch unexpected inconsistencies.
-	if len(blockHeaders) != len(filterHeaders) {
-		return 0, fmt.Errorf("mismatch between block headers "+
-			"(%d) and filter headers (%d)", len(blockHeaders),
-			len(filterHeaders))
-	}
+		// Convert filter header batches to target store format.
+		filterHeaders = make(
+			[]headerfs.FilterHeader, 0, len(filterBatch),
+		)
+		for _, header := range filterBatch {
+			fHeader, err := assertFilterHeader(header)
+			if err != nil {
+				return 0, err
+			}
+			filterHeaders = append(
+				filterHeaders, fHeader.FilterHeader,
+			)
+		}
 
-	if len(blockHeaders) == 0 {
-		return 0, errors.New("no headers read")
+		batchEnd = batchStart + uint32(len(filterBatch)) - 1
```

Comment:
So `batchEnd` is computed twice, one here, and the above, but using the block then filter header length. This is then passed into `writeHeadersToTargetStores`. Looking at the function, it's only used for error logging, but will print inconsistent errors as is. 

To remedy this, I think we need a batch end variable for each header type.